src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -4592,6 +4592,41 @@ public override BoundNode VisitThrowExpression(BoundThrowExpression node)
             return result;
         }
 
+        public override BoundNode VisitYieldReturnStatement(BoundYieldReturnStatement node)
+        {
+            BoundExpression expr = node.Expression;
+            if (expr == null)
+            {
+                return null;
+            }
+            var method = (MethodSymbol)_member;
+            Debug.Assert(method.IsIterator);
+            VisitOptionalImplicitConversion(expr, getIteratorElementType(method.ReturnType.TypeSymbol), useLegacyWarnings: false, AssignmentKind.Return);
+            return null;
+
+            TypeSymbolWithAnnotations getIteratorElementType(TypeSymbol returnType)
+            {
+                switch (returnType.OriginalDefinition.SpecialType)
+                {
+                    case SpecialType.System_Collections_IEnumerable:
+                    case SpecialType.System_Collections_IEnumerator:
+                        return TypeSymbolWithAnnotations.Create(compilation.GetSpecialType(SpecialType.System_Object));
+
+                    case SpecialType.System_Collections_Generic_IEnumerable_T:
+                    case SpecialType.System_Collections_Generic_IEnumerator_T:
+                        return ((NamedTypeSymbol)returnType).TypeArgumentsNoUseSiteDiagnostics[0];
+                }
+
+                if (returnType.OriginalDefinition == compilation.GetWellKnownType(WellKnownType.System_Collections_Generic_IAsyncEnumerable_T))
```

Comment:
>System_Collections_Generic_IAsyncEnumerable_T [](start = 96, length = 45)

More general question: do we want to make this iterator type a SpecialType, like the other iterators? And does async enumerable have an equivalent enumerator type we need to worry about here?

Reply:
From my understanding, "special type" is like a "well-know type", except that it must live in the same assembly as `object` (ie. in the core library).

Yes, there is a corresponding `IAsyncEnumerator<T>` type. Async-iterator methods can't yet return such enumerators yet. That will be in a PR later this week ([https://github.com/dotnet/roslyn/pull/31114](https://github.com/dotnet/roslyn/pull/31114)). I added a note there to update this code too. Good catch

---
In reply to: [232775017](https://github.com/dotnet/roslyn/pull/31072#discussion_r232775017) [](ancestors = 232775017)